src/lib/merkle_ledger/database.ml

```diff
@@ -700,57 +700,86 @@ module Make (Inputs : Inputs_intf) :
     List.map2_exn dependency_dirs dependency_hashes ~f:(fun dir hash ->
         Direction.map dir ~left:(`Left hash) ~right:(`Right hash) )
 
-  let merkle_path_batch mdb locations =
+  let path_batch_impl ~update_locs ~extract_hashes_exn mdb locations =
```

Comment:
I got a version of `path_batch_impl` that looks like the following:
```
  let path_batch_impl ~is_wide mdb locations =
    let locations =
      List.map locations ~f:(fun location ->
          if Location.is_account location then
            Location.Hash (Location.to_path_exn location)
          else (
            assert (Location.is_hash location) ;
            location ) )
    in
    let list_of_dependencies =
      List.map locations ~f:Location.merkle_path_dependencies_exn
    in
    let all_locs =
      let dependency_locs =
        List.concat list_of_dependencies |> List.map ~f:fst
      in
      if is_wide then
        dependency_locs @ List.map dependency_locs ~f:Location.sibling
      else dependency_locs
    in
    let hashes = get_hash_batch_exn mdb all_locs in
    let rec go list_of_dependencies sibling_hashes self_hashes acc =
      match list_of_dependencies with
      | [] ->
          acc
      | deps :: rest_of_deps ->
          let locs, dirs = List.unzip deps in
          let length_of_locs = List.length locs in
          let sibling_hashes, rest_of_sibling_hashes =
            List.(split_n sibling_hashes length_of_locs)
          in
          let self_hashes, rest_of_self_hashes =
            List.(split_n self_hashes length_of_locs)
          in
          let path =
            List.map3_exn dirs sibling_hashes self_hashes
              ~f:(fun dir sibling_hash self_hash ->
                Direction.map dir
                  ~left:
                    (`Left
                      ( if is_wide then (sibling_hash, self_hash)
                      else (sibling_hash, sibling_hash) ) )
                  ~right:
                    (`Right
                      ( if is_wide then (sibling_hash, self_hash)
                      else (sibling_hash, sibling_hash) ) ) )
          in
          go rest_of_deps rest_of_sibling_hashes rest_of_self_hashes
            (path :: acc)
    in
    if is_wide then
      let sibling_hashes, self_hashes =
        List.(split_n hashes (length hashes / 2))
      in
      go list_of_dependencies sibling_hashes self_hashes []
    else go list_of_dependencies hashes hashes []

  let merkle_path_batch md locs =
    path_batch_impl ~is_wide:false md locs
    |> List.map ~f:(fun p ->
           List.map p ~f:(function
             | `Left (h, _) ->
                 `Left h
             | `Right (h, _) ->
                 `Right h ) )

  let wide_merkle_path_batch = path_batch_impl ~is_wide:true
```
I would start a separate branch and test it tomorrow.

Reply:
The basic idea is again just collects the locations and compute the hashes at once.
When we compute the hash, we just scan through the `list_of_dependencies` and then pick the corresponding hashes from the list.

Reply:
Change integrated into the branch (after some discussion and experimentation)